lib/Sema/TypeCheckDecl.cpp

```diff
@@ -7808,9 +7808,9 @@ static void validateAttributes(TypeChecker &TC, Decl *D) {
       TC.diagnose(OA->getLocation(),
                   diag::optional_attribute_non_protocol);
       D->getAttrs().removeAttribute(OA);
-    } else if (!cast<ProtocolDecl>(D->getDeclContext())->isObjC()) {
+    } else if (!D->getAttrs().getAttribute<ObjCAttr>()) {
       TC.diagnose(OA->getLocation(),
-                  diag::optional_attribute_non_objc_protocol);
+                  diag::optional_attribute_non_objc_attribute);
       D->getAttrs().removeAttribute(OA);
```

Comment:
The recovery here should be better: there is no reason to drop the `optional` when it's just missing the `@objc`. I suggest leaving the original error in place. Then, add a separate check that complains about a missing, explicit '@objc' and provides a Fix-It, e.g.:

    if (!D->getAttrs().getAttribute<ObjCAttr>()) {
        TC.diagnose(OA->getLocation(), diag::optional_attribute_non_objc_attribute);
          .fixItInsert(D->getAttributeInsertionLoc(false), "@objc ")
    }